From 2087a0dfd6d68c3a986d3888480be1698198dd7d Mon Sep 17 00:00:00 2001
From: ylevavi <yehuda.levavi@tandemg.com>
Date: Wed, 06 Oct 2021 14:38:58 +0300
Subject: [PATCH] [SW-56463] tpm: add tpm driver

Add TPM driver that supports 3 basic operations:
	- tpm2_startup
	- tpm2_nv_read
	- tpm2_pcr_extend

Change-Id: I8c66d6018110ddd04a4666db0ba69b6032e70ecb
Signed-off-by: ylevavi <yehuda.levavi@tandemg.com>
---

diff --git a/arch/gaudi2/common/Makefile b/arch/gaudi2/common/Makefile
index d476c6a..0ad2425 100644
--- a/arch/gaudi2/common/Makefile
+++ b/arch/gaudi2/common/Makefile
@@ -3,6 +3,7 @@
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH_COMMON)/$(HL_TARGET).c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH_COMMON)/pll.c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH_COMMON)/sha.c)
+PREBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH_COMMON)/tpm2.c)
 
 PPBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH_COMMON)/common.c)
 PPBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH_COMMON)/$(HL_TARGET).c)
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 4da94eb..2599d59 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -10,6 +10,7 @@
 #include <psoc_global_conf_regs.h>
 #include <psoc_global_conf_masks.h>
 #include <psoc_gpio1_regs.h>
+#include <psoc_gpio2_regs.h>
 #include <psoc_gpio3_regs.h>
 #include <psoc_gpio5_regs.h>
 #include <psoc_qspi_regs.h>
@@ -708,9 +709,39 @@
 	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
 }
 
+bool tpm_is_supported(void) { return true; }
+
 unsigned int get_cs_pad_sel_flash(void)
 {
 	return mmPSOC_GLOBAL_CONF_PAD_SEL_19;
+}
+
+unsigned int get_miso_pad_sel_tpm(void)
+{
+	return mmPSOC_GLOBAL_CONF_PAD_SEL_40;
+}
+
+unsigned int get_mosi_pad_sel_tpm(void)
+{
+	return mmPSOC_GLOBAL_CONF_PAD_SEL_41;
+}
+
+unsigned int get_sclk_out_pad_sel_tpm(void)
+{
+	return mmPSOC_GLOBAL_CONF_PAD_SEL_42;
+}
+
+unsigned int get_cs_pad_sel_tpm(void)
+{
+	return mmPSOC_GLOBAL_CONF_PAD_SEL_43;
+}
+
+void spi_ssi_cs1_get(unsigned int *mask, unsigned int *data_offset,
+		     unsigned int *out_en_offset)
+{
+	*mask = BIT(43 % 16);
+	*data_offset = mmPSOC_GPIO2_DATAOUT;
+	*out_en_offset = mmPSOC_GPIO2_OUTENSET;
 }
 
 void spi_ssi_cs0_get(unsigned int *mask, unsigned int *data_offset,
@@ -726,6 +757,11 @@
 	*regs = mmPSOC_QSPI_CTRLR0;
 }
 
+void spi_tpm_base_get(unsigned int *regs)
+{
+	*regs = mmPSOC_SPI_CTRLR0;
+}
+
 u32 hl_pll_timestamp_base_get(void)
 {
 	return mmCPU_TIMESTAMP_CNTCR;
diff --git a/arch/gaudi2/common/tpm2.c b/arch/gaudi2/common/tpm2.c
new file mode 100644
index 0000000..ba6fc72
--- /dev/null
+++ b/arch/gaudi2/common/tpm2.c
@@ -0,0 +1,1057 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#include <stddef.h>
+#include <stdarg.h>
+#include <tpm2.h>
+#include <uart.h>
+#include <preboot.h>
+#include <types.h>
+#include <string.h>
+#include <pll.h>
+#include <spi.h>
+#include <errno.h>
+
+#include "lzo/unaligned.h"
+#include "tpm_internal.h"
+
+/**
+ * TPM2 Structure Tags for command/response buffers.
+ *
+ * @TPM2_ST_NO_SESSIONS: the command does not need an authentication.
+ * @TPM2_ST_SESSIONS: the command needs an authentication.
+ */
+enum tpm2_structures {
+	TPM2_ST_NO_SESSIONS	= 0x8001,
+	TPM2_ST_SESSIONS	= 0x8002,
+};
+
+/**
+ * TPM2 command codes used at the beginning of a buffer, gives the command.
+ *
+ * @TPM2_CC_STARTUP: TPM2_Startup().
+ * @TPM2_CC_SELF_TEST: TPM2_SelfTest().
+ * @TPM2_CC_CLEAR: TPM2_Clear().
+ * @TPM2_CC_CLEARCONTROL: TPM2_ClearControl().
+ * @TPM2_CC_HIERCHANGEAUTH: TPM2_HierarchyChangeAuth().
+ * @TPM2_CC_PCR_SETAUTHPOL: TPM2_PCR_SetAuthPolicy().
+ * @TPM2_CC_DAM_RESET: TPM2_DictionaryAttackLockReset().
+ * @TPM2_CC_DAM_PARAMETERS: TPM2_DictionaryAttackParameters().
+ * @TPM2_CC_GET_CAPABILITY: TPM2_GetCapibility().
+ * @TPM2_CC_GET_RANDOM: TPM2_GetRandom().
+ * @TPM2_CC_PCR_READ: TPM2_PCR_Read().
+ * @TPM2_CC_PCR_EXTEND: TPM2_PCR_Extend().
+ * @TPM2_CC_PCR_SETAUTHVAL: TPM2_PCR_SetAuthValue().
+ */
+enum tpm2_command_codes {
+	TPM2_CC_STARTUP		= 0x0144,
+	TPM2_CC_SELF_TEST	= 0x0143,
+	TPM2_CC_CLEAR		= 0x0126,
+	TPM2_CC_CLEARCONTROL	= 0x0127,
+	TPM2_CC_HIERCHANGEAUTH	= 0x0129,
+	TPM2_CC_PCR_SETAUTHPOL	= 0x012C,
+	TPM2_CC_DAM_RESET	= 0x0139,
+	TPM2_CC_DAM_PARAMETERS	= 0x013A,
+	TPM2_CC_NV_READ         = 0x014E,
+	TPM2_CC_GET_CAPABILITY	= 0x017A,
+	TPM2_CC_GET_RANDOM      = 0x017B,
+	TPM2_CC_PCR_READ	= 0x017E,
+	TPM2_CC_PCR_EXTEND	= 0x0182,
+	TPM2_CC_PCR_SETAUTHVAL	= 0x0183,
+};
+
+enum tpm_timeout {
+	TPM_TIMEOUT_MS			= 5,
+	TIS_SHORT_TIMEOUT_MS		= 750,
+	TIS_LONG_TIMEOUT_MS		= 2000,
+	SLEEP_DURATION_US		= 60,
+	SLEEP_DURATION_LONG_US		= 210,
+};
+
+enum {
+	TPM_STS_FAMILY_SHIFT		= 26,
+	TPM_STS_FAMILY_MASK		= 0x3 << TPM_STS_FAMILY_SHIFT,
+	TPM_STS_FAMILY_TPM2		= 1 << TPM_STS_FAMILY_SHIFT,
+	TPM_STS_RESE_TESTABLISMENT_BIT	= 1 << 25,
+	TPM_STS_COMMAND_CANCEL		= 1 << 24,
+	TPM_STS_BURST_COUNT_SHIFT	= 8,
+	TPM_STS_BURST_COUNT_MASK	= 0xffff << TPM_STS_BURST_COUNT_SHIFT,
+	TPM_STS_VALID			= 1 << 7,
+	TPM_STS_COMMAND_READY		= 1 << 6,
+	TPM_STS_GO			= 1 << 5,
+	TPM_STS_DATA_AVAIL		= 1 << 4,
+	TPM_STS_DATA_EXPECT		= 1 << 3,
+	TPM_STS_SELF_TEST_DONE		= 1 << 2,
+	TPM_STS_RESPONSE_RETRY		= 1 << 1,
+};
+
+enum tis_access {
+	TPM_ACCESS_VALID		= 0x80,
+	TPM_ACCESS_ACTIVE_LOCALITY	= 0x20,
+	TPM_ACCESS_REQUEST_PENDING	= 0x04,
+	TPM_ACCESS_REQUEST_USE		= 0x02,
+};
+
+/* Number of wait states to wait for */
+#define TPM_WAIT_STATES 100
+
+#define TPM_ACCESS(l)			(0x0000 | ((l) << 12))
+#define TPM_INT_ENABLE(l)               (0x0008 | ((l) << 12))
+#define TPM_STS(l)			(0x0018 | ((l) << 12))
+#define TPM_DATA_FIFO(l)		(0x0024 | ((l) << 12))
+#define TPM_DID_VID(l)			(0x0F00 | ((l) << 12))
+#define TPM_RID(l)			(0x0F04 | ((l) << 12))
+
+#define MAX_SPI_FRAMESIZE 32
+
+#define COMMAND_BUFFER_SIZE 256
+
+/* Internal error of TPM command library */
+#define TPM_LIB_ERROR ((u32)~0u)
+
+/* To make strings of commands more easily */
+#define __MSB(x) ((x) >> 8)
+#define __LSB(x) ((x) & 0xFF)
+#define tpm_u16(x) __MSB(x), __LSB(x)
+#define tpm_u32(x) tpm_u16((x) >> 16), tpm_u16((x) & 0xFFFF)
+
+#define TPM_HEADER_SIZE		10
+
+/* Max buffer size supported by our tpm */
+#define TPM_DEV_BUFSIZE		1260
+
+#define MS_TO_US(m) ((m) * 1000)
+
+#define min_t(type, x, y) ({			\
+	type __min1 = (x);			\
+	type __min2 = (y);			\
+	__min1 < __min2 ? __min1 : __min2; })
+
+
+#define le32_to_cpu(x) ((__force __u32)(__le32)(x))
+
+/**
+ * TPM2 permanent handles.
+ *
+ * @TPM2_RH_OWNER: refers to the 'owner' hierarchy.
+ * @TPM2_RS_PW: indicates a password.
+ * @TPM2_RH_LOCKOUT: refers to the 'lockout' hierarchy.
+ * @TPM2_RH_ENDORSEMENT: refers to the 'endorsement' hierarchy.
+ * @TPM2_RH_PLATFORM: refers to the 'platform' hierarchy.
+ */
+enum tpm2_handles {
+	TPM2_RH_OWNER		= 0x40000001,
+	TPM2_RS_PW		= 0x40000009,
+	TPM2_RH_LOCKOUT		= 0x4000000A,
+	TPM2_RH_ENDORSEMENT	= 0x4000000B,
+	TPM2_RH_PLATFORM	= 0x4000000C,
+};
+
+enum {
+	/* Secure storage for firmware settings */
+	TPM_HT_PCR = 0,
+	TPM_HT_NV_INDEX,
+	TPM_HT_HMAC_SESSION,
+	TPM_HT_POLICY_SESSION,
+
+	HR_SHIFT		= 24,
+	HR_PCR			= TPM_HT_PCR << HR_SHIFT,
+	HR_HMAC_SESSION		= TPM_HT_HMAC_SESSION << HR_SHIFT,
+	HR_POLICY_SESSION	= TPM_HT_POLICY_SESSION << HR_SHIFT,
+	HR_NV_INDEX		= TPM_HT_NV_INDEX << HR_SHIFT,
+};
+
+static struct spi_ctx tpm_spi = {0};
+static int locality;
+
+__weak bool tpm_is_supported(void) { return false; }
+
+__weak unsigned int get_cs_pad_sel_tpm(void) { return 0; }
+
+__weak unsigned int get_miso_pad_sel_tpm(void) { return 0; }
+
+__weak unsigned int get_mosi_pad_sel_tpm(void) { return 0; }
+
+__weak unsigned int get_sclk_out_pad_sel_tpm(void) { return 0; }
+
+__weak void spi_tpm_base_get(unsigned int *regs) {}
+
+__weak void spi_ssi_cs1_get(unsigned int *mask, unsigned int *data_offset,
+				unsigned int *out_en_offset)
+{}
+
+static void spi_ssi_cs_init(void)
+{
+	unsigned int reg, mask, data, out_en;
+
+	spi_ssi_cs1_get(&mask, &data, &out_en);
+
+	reg  = regmap_readl(data);
+	reg  &= ~mask;
+	regmap_writel(reg, data);
+
+	reg = regmap_readl(out_en);
+	reg |= mask;
+	regmap_writel(reg, out_en);
+
+	regmap_writel(1, get_miso_pad_sel_tpm());
+	regmap_writel(1, get_mosi_pad_sel_tpm());
+	regmap_writel(1, get_sclk_out_pad_sel_tpm());
+}
+
+static void spi_tpm_claim_bus(void)
+{
+	regmap_writel(0, get_cs_pad_sel_tpm());
+}
+
+static void spi_tpm_release_bus(void)
+{
+	regmap_writel(1, get_cs_pad_sel_tpm());
+}
+
+static void tpm_spi_init(void)
+{
+	spi_tpm_base_get(&tpm_spi.spi_base);
+	tpm_spi.spi_initialized = false;
+	tpm_spi.spi_access_is_locked = false;
+	tpm_spi.cs_init = spi_ssi_cs_init;
+	tpm_spi.claim_bus = spi_tpm_claim_bus;
+	tpm_spi.release_bus = spi_tpm_release_bus;
+	tpm_spi.fifo_len = 0;
+	spi_init(&tpm_spi);
+}
+
+static inline int tpm_command_size(const void *command)
+{
+	const size_t command_size_offset = 2;
+
+	/* get unaligned be32 */
+	return get_unaligned_be32((__be32 *)(command + command_size_offset));
+}
+
+static u32 tpm_return_code(const void *response)
+{
+	const size_t return_code_offset = 6;
+
+	return get_unaligned_be32(response + return_code_offset);
+}
+
+/* Returns max number of milliseconds to wait */
+static u64 tpm_tis_calc_ordinal_duration(u32 ordinal)
+{
+	int duration_idx = TPM_UNDEFINED;
+	int duration = 0;
+
+	duration_ms[TPM_SHORT] = TIS_SHORT_TIMEOUT_MS;
+	duration_ms[TPM_MEDIUM] = TIS_LONG_TIMEOUT_MS;
+	duration_ms[TPM_LONG] = TIS_LONG_TIMEOUT_MS;
+
+	if (ordinal < TPM_MAX_ORDINAL) {
+		duration_idx = tpm_ordinal_duration[ordinal];
+	} else if ((ordinal & TPM_PROTECTED_ORDINAL_MASK) <
+			TPM_MAX_PROTECTED_ORDINAL) {
+		duration_idx = tpm_protected_ordinal_duration[
+				ordinal & TPM_PROTECTED_ORDINAL_MASK];
+	}
+
+	if (duration_idx != TPM_UNDEFINED)
+		duration = duration_ms[duration_idx];
+
+	if (duration <= 0)
+		return 2 * 60 * 1000; /* Two minutes timeout */
+	else
+		return duration;
+}
+
+/**
+ * tpm_tis_spi_read() - Read from TPM register
+ *
+ * @addr: register address to read from
+ * @buffer: provided by caller
+ * @len: number of bytes to read
+ *
+ * Read len bytes from TPM register and put them into
+ * buffer (little-endian format, i.e. first byte is put into buffer[0]).
+ *
+ * NOTE: TPM is big-endian for multi-byte values. Multi-byte
+ * values have to be swapped.
+ *
+ * @return -EIO on error, 0 on success.
+ */
+static int tpm_tis_spi_xfer(u32 addr, const u8 *out, u8 *in, u16 len)
+{
+	int transfer_len, ret;
+	u8 tx_buf[MAX_SPI_FRAMESIZE];
+	u8 rx_buf[MAX_SPI_FRAMESIZE];
+
+	if (in && out) {
+		preboot_puts("TPM2: can't do full duplex\n");
+		return -EINVAL;
+	}
+
+	spi_tpm_claim_bus();
+
+	while (len) {
+		/* Request */
+		transfer_len = min_t(u16, len, MAX_SPI_FRAMESIZE);
+		tx_buf[0] = (in ? BIT(7) : 0) | (transfer_len - 1);
+		tx_buf[1] = 0xD4;
+		tx_buf[2] = addr >> 8;
+		tx_buf[3] = addr;
+
+		ret = spi_xfer(&tpm_spi, 4 * 8, tx_buf, rx_buf, SPI_XFER_BEGIN);
+		if (ret < 0) {
+			preboot_puts("TPM2: spi request transfer failed\n");
+			goto release_bus;
+		}
+
+		/* Wait state */
+		if (!(rx_buf[3] & 0x1)) {
+			int i;
+
+			for (i = 0; i < TPM_WAIT_STATES; i++) {
+				ret = spi_xfer(&tpm_spi, 1 * 8, NULL, rx_buf,
+						0);
+				if (ret) {
+					preboot_puts_hex("TPM2: wait state failed: ",
+								ret, true);
+					goto release_bus;
+				}
+
+				if (rx_buf[0] & 0x1)
+					break;
+			}
+
+			if (i == TPM_WAIT_STATES) {
+				preboot_puts("TPM2: timeout on wait state\n");
+				ret = -1;
+				goto release_bus;
+			}
+		}
+
+		/* Read/Write */
+		if (out) {
+			memcpy(tx_buf, out, transfer_len);
+			out += transfer_len;
+		}
+
+		ret = spi_xfer(&tpm_spi, transfer_len * 8,
+			       out ? tx_buf : NULL,
+			       in ? rx_buf : NULL,
+			       SPI_XFER_END);
+		if (ret) {
+			preboot_puts("TPM2: spi read transfer failed\n");
+			goto release_bus;
+		}
+
+		if (in) {
+			memcpy(in, rx_buf, transfer_len);
+			in += transfer_len;
+		}
+
+		len -= transfer_len;
+	}
+
+release_bus:
+	/* If an error occurred, release the chip by deasserting the CS */
+	if (ret < 0)
+		spi_xfer(&tpm_spi, 0, NULL, NULL, SPI_XFER_END);
+
+	spi_tpm_release_bus();
+
+	return ret;
+}
+
+static int tpm_tis_spi_read(u16 addr, u8 *in, u16 len)
+{
+	return tpm_tis_spi_xfer(addr, NULL, in, len);
+}
+
+static int tpm_tis_spi_read32(u32 addr, u32 *result)
+{
+	__le32 result_le;
+	int ret;
+
+	ret = tpm_tis_spi_read(addr, (u8 *)&result_le, sizeof(u32));
+	if (!ret)
+		*result = le32_to_cpu(result_le);
+
+	return ret;
+}
+
+static u8 tpm_spi_status(u8 *status)
+{
+	return tpm_tis_spi_read(TPM_STS(locality), status, 1);
+}
+
+static int tpm_spi_wait_for_stat(u8 mask, unsigned long timeout, u8 *status)
+{
+	u64 start = timestamp_get_us();
+	u64 stop = timeout;
+	int ret;
+
+	do {
+		mdelay(TPM_TIMEOUT_MS);
+		ret = tpm_spi_status(status);
+		if (ret)
+			return ret;
+
+		if ((*status & mask) == mask)
+			return 0;
+	} while (!timeout_expired(start, MS_TO_US(stop)));
+
+	return -1;
+}
+
+static int tpm_spi_get_burstcount(void)
+{
+	u64 start, stop;
+	u32 burstcount, ret;
+
+	/* wait for burstcount */
+	start = timestamp_get_us();
+	stop = TIS_SHORT_TIMEOUT_MS;
+	do {
+		ret = tpm_tis_spi_read32(TPM_STS(locality),
+					 &burstcount);
+		if (ret)
+			return -1;
+
+		burstcount = (burstcount >> 8) & 0xFFFF;
+		if (burstcount)
+			return burstcount;
+
+		mdelay(TPM_TIMEOUT_MS);
+	} while (!timeout_expired(start, MS_TO_US(stop)));
+
+	return -1;
+}
+
+static int tpm_spi_recv_data(u8 *buf, size_t count)
+{
+	int size = 0, burstcnt, len, ret;
+	u8 status;
+
+	while (size < count &&
+	       tpm_spi_wait_for_stat(TPM_STS_DATA_AVAIL | TPM_STS_VALID,
+					 TIS_SHORT_TIMEOUT_MS, &status) == 0) {
+		burstcnt = tpm_spi_get_burstcount();
+		if (burstcnt < 0)
+			return burstcnt;
+
+		len = min_t(int, burstcnt, count - size);
+		ret = tpm_tis_spi_read(TPM_DATA_FIFO(locality),
+				       buf + size, len);
+		if (ret < 0)
+			return ret;
+
+		size += len;
+	}
+
+	return size;
+}
+
+static int tpm_tis_spi_write(u16 addr, const u8 *out, u16 len)
+{
+	return tpm_tis_spi_xfer(addr, out, NULL, len);
+}
+
+static int tpm_tis_spi_cancel(void)
+{
+	u8 data = TPM_STS_COMMAND_READY;
+
+	return tpm_tis_spi_write(TPM_STS(locality), &data, 1);
+}
+
+static void tpm_tis_spi_release_locality(int loc, bool force)
+{
+	const u8 mask = TPM_ACCESS_REQUEST_PENDING | TPM_ACCESS_VALID;
+	u8 buf;
+
+	if (tpm_tis_spi_read(TPM_ACCESS(loc), &buf, 1) < 0)
+		return;
+
+	if (force || (buf & mask) == mask) {
+		buf = TPM_ACCESS_ACTIVE_LOCALITY;
+		tpm_tis_spi_write(TPM_ACCESS(loc), &buf, 1);
+	}
+}
+
+static int tpm_spi_recv(u8 *buf, size_t count)
+{
+	int size, expected;
+
+	if (count < TPM_HEADER_SIZE) {
+		size = -1;
+		goto out;
+	}
+
+	size = tpm_spi_recv_data(buf, TPM_HEADER_SIZE);
+	if (size < TPM_HEADER_SIZE) {
+		preboot_puts("TPM error, unable to read header\n");
+		goto out;
+	}
+
+	expected = get_unaligned_be32(buf + 2);
+	if (expected > count) {
+		size = -1;
+		goto out;
+	}
+
+	size += tpm_spi_recv_data(&buf[TPM_HEADER_SIZE],
+				   expected - TPM_HEADER_SIZE);
+	if (size < expected) {
+		preboot_puts("TPM error, unable to read remaining bytes of result\n");
+		size = -1;
+		goto out;
+	}
+
+out:
+	tpm_tis_spi_cancel();
+	tpm_tis_spi_release_locality(locality, false);
+
+	return size;
+}
+
+static int tpm_tis_spi_check_locality(int loc)
+{
+	const u8 mask = TPM_ACCESS_ACTIVE_LOCALITY | TPM_ACCESS_VALID;
+	u8 buf;
+	int ret;
+
+	ret = tpm_tis_spi_read(TPM_ACCESS(loc), &buf, 1);
+	if (ret)
+		return ret;
+
+	if ((buf & mask) == mask) {
+		locality = loc;
+		return 0;
+	}
+
+	return -ENOENT;
+}
+
+static int tpm_tis_spi_request_locality(int loc)
+{
+	unsigned long start, stop;
+	u8 buf = TPM_ACCESS_REQUEST_USE;
+	int ret;
+
+	ret = tpm_tis_spi_check_locality(loc);
+	if (!ret)
+		return 0;
+
+	if (ret != -ENOENT) {
+		preboot_puts("TPM2: Failed to get locality\n");
+		return ret;
+	}
+
+	ret = tpm_tis_spi_write(TPM_ACCESS(loc), &buf, 1);
+	if (ret) {
+		preboot_puts("TPM2: Failed to write to TPM\n");
+		return ret;
+	}
+
+	start = timestamp_get_us();
+	stop = TIS_SHORT_TIMEOUT_MS;
+	do {
+		ret = tpm_tis_spi_check_locality(loc);
+		if (!ret)
+			return 0;
+
+		if (ret != -ENOENT) {
+			preboot_puts("TPM2: Failed to get locality\n");
+			return ret;
+		}
+
+		mdelay(TPM_TIMEOUT_MS);
+	} while (!timeout_expired(start, MS_TO_US(stop)));
+
+	preboot_puts("TPM2: Timeout getting locality\n");
+
+	return ret;
+}
+
+static u8 tpm_tis_spi_valid_status(u8 *status)
+{
+	return tpm_spi_wait_for_stat(TPM_STS_VALID, TIS_SHORT_TIMEOUT_MS,
+					status);
+}
+
+static int tpm_spi_send(const u8 *buf, size_t len)
+{
+	u32 i, size;
+	u8 status;
+	int burstcnt, ret;
+	u8 data;
+
+	if (len > TPM_DEV_BUFSIZE)
+		return -E2BIG;  /* Command is too long for our tpm, sorry */
+
+	ret = tpm_tis_spi_request_locality(0);
+	if (ret < 0)
+		return -EBUSY;
+
+	/*
+	 * Check if the TPM is ready. If not, if not, cancel the pending command
+	 * and poll on the status to be finally ready.
+	 */
+	ret = tpm_spi_status(&status);
+	if (ret)
+		return ret;
+
+	if (!(status & TPM_STS_COMMAND_READY)) {
+		/* Force the transition, usually this will be done at startup */
+		ret = tpm_tis_spi_cancel();
+		if (ret) {
+			preboot_puts("TPM2: Could not cancel previous operation\n");
+			goto out_err;
+		}
+
+		ret = tpm_spi_wait_for_stat(TPM_STS_COMMAND_READY,
+						TIS_LONG_TIMEOUT_MS, &status);
+		if (ret < 0 || !(status & TPM_STS_COMMAND_READY)) {
+			preboot_puts("status not ready after wait for stat returned err\n");
+			goto out_err;
+		}
+	}
+
+	for (i = 0; i < len - 1;) {
+		burstcnt = tpm_spi_get_burstcount();
+		if (burstcnt < 0)
+			return burstcnt;
+
+		size = min_t(int, len - i - 1, burstcnt);
+		ret = tpm_tis_spi_write(TPM_DATA_FIFO(locality),
+					buf + i, size);
+		if (ret < 0)
+			goto out_err;
+
+		i += size;
+	}
+
+	ret = tpm_tis_spi_valid_status(&status);
+	if (ret)
+		goto out_err;
+
+	if ((status & TPM_STS_DATA_EXPECT) == 0) {
+		ret = -EIO;
+		goto out_err;
+	}
+
+	ret = tpm_tis_spi_write(TPM_DATA_FIFO(locality),
+				buf + len - 1, 1);
+	if (ret)
+		goto out_err;
+
+	ret = tpm_tis_spi_valid_status(&status);
+	if (ret)
+		goto out_err;
+
+	if ((status & TPM_STS_DATA_EXPECT) != 0) {
+		ret = -EIO;
+		goto out_err;
+	}
+
+	data = TPM_STS_GO;
+	ret = tpm_tis_spi_write(TPM_STS(locality), &data, 1);
+	if (ret)
+		goto out_err;
+
+	return len;
+
+out_err:
+	tpm_tis_spi_cancel();
+	tpm_tis_spi_release_locality(locality, false);
+
+	return ret;
+}
+
+static int tpm_spi_cleanup(void)
+{
+	tpm_tis_spi_cancel();
+	/*
+	 * The TPM needs some time to clean up here,
+	 * so we sleep rather than keeping the bus busy
+	 */
+	mdelay(2);
+	tpm_tis_spi_release_locality(locality, false);
+
+	return 0;
+}
+
+static int tpm_xfer(const uint8_t *sendbuf, size_t send_size,
+	uint8_t *recvbuf, size_t *recv_size)
+{
+	u64 start, stop;
+	unsigned int count, ordinal;
+	int ret = 0;
+	u8 buf[TPM_DEV_BUFSIZE + sizeof(u8)];  /* Max buffer size + addr */
+
+	/* switch endianness: big->little */
+	count = get_unaligned_be32(sendbuf + TPM_CMD_COUNT_BYTE);
+	ordinal = get_unaligned_be32(sendbuf + TPM_CMD_ORDINAL_BYTE);
+
+	if (count == 0) {
+		preboot_puts("TPM2: no data\n");
+		return -1;
+	}
+	if (count > send_size) {
+		preboot_puts("TPM2: invalid count value\n");
+		return -1;
+	}
+
+	ret = tpm_spi_send(sendbuf, send_size);
+	if (ret < 0)
+		return ret;
+
+	start = timestamp_get_us();
+	stop = tpm_tis_calc_ordinal_duration(ordinal);
+	do {
+		ret = tpm_spi_recv(buf, sizeof(buf));
+		if (ret >= 0) {
+			if (ret > *recv_size)
+				return -1;
+			memcpy(recvbuf, buf, ret);
+			*recv_size = ret;
+			ret = 0;
+			break;
+		} else if (ret != -1) {
+			return ret;
+		}
+
+		mdelay(TPM_TIMEOUT_MS);
+		if (timeout_expired(start, MS_TO_US(stop))) {
+			preboot_puts("TPM2: timeout expired\n");
+			ret = -1;
+			break;
+		}
+	} while (ret);
+
+	if (ret)
+		tpm_spi_cleanup();
+
+	return 0;
+}
+
+static int tpm_spi_sendrecv_command(const void *command,
+			 void *response, size_t *size_ptr)
+{
+	int err, ret;
+	u8 response_buffer[COMMAND_BUFFER_SIZE];
+	size_t response_length;
+
+	if (response) {
+		response_length = *size_ptr;
+	} else {
+		response = response_buffer;
+		response_length = sizeof(response_buffer);
+	}
+
+	err = tpm_xfer(command, tpm_command_size(command),
+		       response, &response_length);
+
+	if (err < 0)
+		return err;
+
+	if (size_ptr)
+		*size_ptr = response_length;
+
+	ret = tpm_return_code(response);
+
+	return ret;
+}
+
+int tpm2_startup(enum tpm2_startup_types mode)
+{
+	const u8 command_v2[12] = {
+		tpm_u16(TPM2_ST_NO_SESSIONS),
+		tpm_u32(12),
+		tpm_u32(TPM2_CC_STARTUP),
+		tpm_u16(mode),
+	};
+	int ret;
+
+	/*
+	 * Note TPM2_Startup command will return RC_SUCCESS the first time,
+	 * but will return RC_INITIALIZE otherwise.
+	 */
+	ret = tpm_spi_sendrecv_command(command_v2, NULL, NULL);
+	if (ret && ret != TPM2_RC_INITIALIZE)
+		return ret;
+
+	return 0;
+}
+
+static int tpm2_unpack_byte_string(const u8 *str, size_t size,
+					const char *format, ...)
+{
+	va_list args;
+	size_t offset = 0, length = 0;
+	u8 *ptr8 = NULL;
+	u16 *ptr16 = NULL;
+	u32 *ptr32 = NULL;
+
+	va_start(args, format);
+	for (; *format; format++) {
+		switch (*format) {
+		case 'b':
+			offset = va_arg(args, size_t);
+			ptr8 = va_arg(args, u8 *);
+			length = 1;
+			break;
+		case 'w':
+			offset = va_arg(args, size_t);
+			ptr16 = va_arg(args, u16 *);
+			length = 2;
+			break;
+		case 'd':
+			offset = va_arg(args, size_t);
+			ptr32 = va_arg(args, u32 *);
+			length = 4;
+			break;
+		case 's':
+			offset = va_arg(args, size_t);
+			ptr8 = va_arg(args, u8 *);
+			length = va_arg(args, u32);
+			break;
+		default:
+			va_end(args);
+			preboot_puts("TPM2: Couldn't recognize format string\n");
+			return -1;
+		}
+
+		if (offset + length > size) {
+			va_end(args);
+			preboot_puts("TPM2: Failed to read\n");
+			return -1;
+		}
+
+		switch (*format) {
+		case 'b':
+			*ptr8 = str[offset];
+			break;
+		case 'w':
+			*ptr16 = get_unaligned_be16(str + offset);
+			break;
+		case 'd':
+			*ptr32 = get_unaligned_be32(str + offset);
+			break;
+		case 's':
+			memcpy(ptr8, str + offset, length);
+			break;
+		}
+	}
+	va_end(args);
+
+	return 0;
+}
+
+static int tpm2_pack_byte_string(u8 *str, size_t size, const char *format, ...)
+{
+	va_list args;
+	size_t offset = 0, length = 0;
+	u8 *data = NULL;
+	u32 value = 0;
+
+	va_start(args, format);
+	for (; *format; format++) {
+		switch (*format) {
+		case 'b':
+			offset = va_arg(args, size_t);
+			value = va_arg(args, int);
+			length = 1;
+			break;
+		case 'w':
+			offset = va_arg(args, size_t);
+			value = va_arg(args, int);
+			length = 2;
+			break;
+		case 'd':
+			offset = va_arg(args, size_t);
+			value = va_arg(args, u32);
+			length = 4;
+			break;
+		case 's':
+			offset = va_arg(args, size_t);
+			data = va_arg(args, u8 *);
+			length = va_arg(args, u32);
+			break;
+		default:
+			preboot_puts("TPM2: Couldn't recognize format string\n");
+			va_end(args);
+			return -1;
+		}
+
+		if (offset + length > size) {
+			va_end(args);
+			return -1;
+		}
+
+		switch (*format) {
+		case 'b':
+			str[offset] = value;
+			break;
+		case 'w':
+			put_unaligned_be16(value, str + offset);
+			break;
+		case 'd':
+			put_unaligned_be32(value, str + offset);
+			break;
+		case 's':
+			memcpy(str + offset, data, length);
+			break;
+		}
+	}
+	va_end(args);
+
+	return 0;
+}
+
+int tpm2_nv_read(u32 index, void *data, u32 count)
+{
+	u8 command_v2[COMMAND_BUFFER_SIZE] = {
+		/* header 10 bytes */
+		tpm_u16(TPM2_ST_SESSIONS),	/* TAG */
+		tpm_u32(10 + 8 + 4 + 9 + 4),	/* Length */
+		tpm_u32(TPM2_CC_NV_READ),	/* Command code */
+
+		/* handles 8 bytes */
+		tpm_u32(TPM2_RH_PLATFORM),	/* Primary platform seed */
+		tpm_u32(index),	/* Password authorisation */
+
+		/* AUTH_SESSION */
+		tpm_u32(9),			/* Authorization size */
+		tpm_u32(TPM2_RS_PW),		/* Session handle */
+		tpm_u16(0),			/* Size of <nonce> */
+						/* <nonce> (if any) */
+		0,				/* Attributes: Cont/Excl/Rst */
+		tpm_u16(0),			/* Size of <hmac/password> */
+						/* <hmac/password> (if any) */
+
+		tpm_u16(count),			/* Number of bytes */
+		tpm_u16(0),			/* Offset */
+	};
+	size_t response_len = COMMAND_BUFFER_SIZE;
+	u8 response[COMMAND_BUFFER_SIZE];
+	int ret;
+	u16 tag;
+	u32 size, code;
+
+	ret = tpm_spi_sendrecv_command(command_v2, response, &response_len);
+	if (ret)
+		return ret;
+	if (tpm2_unpack_byte_string(response, response_len, "wdds", 0, &tag, 2,
+					&size, 6, &code, 16, data, count))
+		return TPM_LIB_ERROR;
+
+	return 0;
+}
+
+int tpm2_pcr_extend(u32 index, u32 algorithm, const u8 *digest, u32 digest_len)
+{
+	/* Length of the message header, up to start of digest */
+	u32 offset = 33;
+	u8 command_v2[COMMAND_BUFFER_SIZE] = {
+		tpm_u16(TPM2_ST_SESSIONS),	/* TAG */
+		tpm_u32(offset + digest_len),	/* Length */
+		tpm_u32(TPM2_CC_PCR_EXTEND),	/* Command code */
+
+		/* HANDLE */
+		tpm_u32(index),			/* Handle (PCR Index) */
+
+		/* AUTH_SESSION */
+		tpm_u32(9),			/* Authorization size */
+		tpm_u32(TPM2_RS_PW),		/* Session handle */
+		tpm_u16(0),			/* Size of <nonce> */
+						/* <nonce> (if any) */
+		0,				/* Attributes: Cont/Excl/Rst */
+		tpm_u16(0),			/* Size of <hmac/password> */
+						/* <hmac/password> (if any) */
+
+		/* hashes */
+		tpm_u32(1),			/* Count (number of hashes) */
+		tpm_u16(algorithm),	/* Algorithm of the hash */
+		/* STRING(digest)		   Digest */
+	};
+	int ret;
+
+	/*
+	 * Fill the command structure starting from the first buffer:
+	 *     - the digest
+	 */
+	ret = tpm2_pack_byte_string(command_v2, sizeof(command_v2), "s",
+			       offset, digest, digest_len);
+	if (ret)
+		return TPM_LIB_ERROR;
+
+	return tpm_spi_sendrecv_command(command_v2, NULL, NULL);
+}
+
+static int tpm_tis_wait_init(int loc)
+{
+	unsigned long start, stop;
+	u8 status;
+	int ret;
+
+	start = timestamp_get_us();
+	stop = TIS_LONG_TIMEOUT_MS;
+	do {
+		mdelay(TPM_TIMEOUT_MS);
+
+		ret = tpm_tis_spi_read(TPM_ACCESS(loc), &status, 1);
+		if (ret)
+			break;
+
+		if (status & TPM_ACCESS_VALID)
+			return 0;
+	} while (!timeout_expired(start, MS_TO_US(stop)));
+
+	return -EIO;
+}
+
+int tpm2_init(void)
+{
+	int ret;
+	u32 vend_dev;
+	u8 rid;
+
+	locality = 0;
+
+	if (!tpm_is_supported())
+		return 0;
+
+	tpm_spi_init();
+
+	ret = tpm_tis_wait_init(locality);
+	if (ret) {
+		preboot_puts("TPM2: no device found\n");
+		return ret;
+	}
+
+	ret = tpm_tis_spi_request_locality(locality);
+	if (ret) {
+		preboot_puts("TPM2: could not request locality\n");
+		return ret;
+	}
+
+	ret = tpm_tis_spi_read32(TPM_DID_VID(locality), &vend_dev);
+	if (ret) {
+		preboot_puts("TPM2: could not retrieve VendorID/DeviceID\n");
+		return ret;
+	}
+
+	ret = tpm_tis_spi_read(TPM_RID(locality), &rid, 1);
+	if (ret) {
+		preboot_puts("TPM2: could not retrieve RevisionID\n");
+		return ret;
+	}
+
+	preboot_puts("SPI TPMv2.0 found\n");
+
+	return 0;
+}
diff --git a/arch/gaudi2/common/tpm2.h b/arch/gaudi2/common/tpm2.h
new file mode 100644
index 0000000..33a8730
--- /dev/null
+++ b/arch/gaudi2/common/tpm2.h
@@ -0,0 +1,67 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef _HL_TPM2_H_
+#define _HL_TPM2_H_
+
+#include <types.h>
+#include <stdbool.h>
+
+/**
+ * TPM2 startup values.
+ *
+ * @TPM2_SU_CLEAR: reset the internal state.
+ * @TPM2_SU_STATE: restore saved state (if any).
+ */
+enum tpm2_startup_types {
+	TPM2_SU_CLEAR		= 0x0000,
+	TPM2_SU_STATE		= 0x0001,
+};
+
+/**
+ * TPM2 return codes.
+ */
+enum tpm2_return_codes {
+	TPM2_RC_SUCCESS		= 0x0000,
+	TPM2_RC_BAD_TAG		= 0x001E,
+	TPM2_RC_FMT1		= 0x0080,
+	TPM2_RC_HASH		= TPM2_RC_FMT1 + 0x0003,
+	TPM2_RC_VALUE		= TPM2_RC_FMT1 + 0x0004,
+	TPM2_RC_SIZE		= TPM2_RC_FMT1 + 0x0015,
+	TPM2_RC_BAD_AUTH	= TPM2_RC_FMT1 + 0x0022,
+	TPM2_RC_HANDLE		= TPM2_RC_FMT1 + 0x000B,
+	TPM2_RC_VER1		= 0x0100,
+	TPM2_RC_INITIALIZE	= TPM2_RC_VER1 + 0x0000,
+	TPM2_RC_FAILURE		= TPM2_RC_VER1 + 0x0001,
+	TPM2_RC_DISABLED	= TPM2_RC_VER1 + 0x0020,
+	TPM2_RC_AUTH_MISSING	= TPM2_RC_VER1 + 0x0025,
+	TPM2_RC_COMMAND_CODE	= TPM2_RC_VER1 + 0x0043,
+	TPM2_RC_AUTHSIZE	= TPM2_RC_VER1 + 0x0044,
+	TPM2_RC_AUTH_CONTEXT	= TPM2_RC_VER1 + 0x0045,
+	TPM2_RC_NEEDS_TEST	= TPM2_RC_VER1 + 0x0053,
+	TPM2_RC_WARN		= 0x0900,
+	TPM2_RC_TESTING		= TPM2_RC_WARN + 0x000A,
+	TPM2_RC_REFERENCE_H0	= TPM2_RC_WARN + 0x0010,
+	TPM2_RC_LOCKOUT		= TPM2_RC_WARN + 0x0021,
+};
+
+int tpm2_init(void);
+int tpm2_startup(enum tpm2_startup_types mode);
+int tpm2_nv_read(u32 index, void *data, u32 count);
+int tpm2_pcr_extend(u32 index, u32 algorithm, const u8 *digest, u32 digest_len);
+
+/* implemented per arch */
+bool tpm_is_supported(void);
+void spi_tpm_base_get(unsigned int *regs);
+unsigned int get_cs_pad_sel_tpm(void);
+unsigned int get_miso_pad_sel_tpm(void);
+unsigned int get_mosi_pad_sel_tpm(void);
+unsigned int get_sclk_out_pad_sel_tpm(void);
+void spi_ssi_cs1_get(unsigned int *mask, unsigned int *data_offset,
+		     unsigned int *out_en_offset);
+
+#endif /* _HL_TPM2_H_ */
diff --git a/arch/gaudi2/common/tpm_internal.h b/arch/gaudi2/common/tpm_internal.h
new file mode 100644
index 0000000..b03a86b
--- /dev/null
+++ b/arch/gaudi2/common/tpm_internal.h
@@ -0,0 +1,297 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Copyright (c) 2015 Google, Inc
+ */
+
+#ifndef __tpm_internal_h
+#define __tpm_internal_h
+
+enum {
+	TPM_MAX_ORDINAL			= 243,
+	TPM_MAX_PROTECTED_ORDINAL	= 12,
+	TPM_PROTECTED_ORDINAL_MASK	= 0xff,
+	TPM_CMD_COUNT_BYTE		= 2,
+	TPM_CMD_ORDINAL_BYTE		= 6,
+};
+
+enum tpm_duration {
+	TPM_SHORT = 0,
+	TPM_MEDIUM = 1,
+	TPM_LONG = 2,
+	TPM_UNDEFINED,
+
+	TPM_DURATION_COUNT,
+};
+
+static unsigned int duration_ms[TPM_DURATION_COUNT];
+
+/*
+ * Array with one entry per ordinal defining the maximum amount
+ * of time the chip could take to return the result.  The ordinal
+ * designation of short, medium or long is defined in a table in
+ * TCG Specification TPM Main Part 2 TPM Structures Section 17. The
+ * values of the SHORT, MEDIUM, and LONG durations are retrieved
+ * from the chip during initialization with a call to tpm_get_timeouts.
+ */
+static const u8 tpm_protected_ordinal_duration[TPM_MAX_PROTECTED_ORDINAL] = {
+	TPM_UNDEFINED,		/* 0 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 5 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 10 */
+	TPM_SHORT,
+};
+
+static const u8 tpm_ordinal_duration[TPM_MAX_ORDINAL] = {
+	TPM_UNDEFINED,		/* 0 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 5 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 10 */
+	TPM_SHORT,
+	TPM_MEDIUM,
+	TPM_LONG,
+	TPM_LONG,
+	TPM_MEDIUM,		/* 15 */
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_MEDIUM,
+	TPM_LONG,
+	TPM_SHORT,		/* 20 */
+	TPM_SHORT,
+	TPM_MEDIUM,
+	TPM_MEDIUM,
+	TPM_MEDIUM,
+	TPM_SHORT,		/* 25 */
+	TPM_SHORT,
+	TPM_MEDIUM,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_MEDIUM,		/* 30 */
+	TPM_LONG,
+	TPM_MEDIUM,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,		/* 35 */
+	TPM_MEDIUM,
+	TPM_MEDIUM,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_MEDIUM,		/* 40 */
+	TPM_LONG,
+	TPM_MEDIUM,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,		/* 45 */
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_LONG,
+	TPM_MEDIUM,		/* 50 */
+	TPM_MEDIUM,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 55 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_MEDIUM,		/* 60 */
+	TPM_MEDIUM,
+	TPM_MEDIUM,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_MEDIUM,		/* 65 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 70 */
+	TPM_SHORT,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 75 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_LONG,		/* 80 */
+	TPM_UNDEFINED,
+	TPM_MEDIUM,
+	TPM_LONG,
+	TPM_SHORT,
+	TPM_UNDEFINED,		/* 85 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 90 */
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_UNDEFINED,		/* 95 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_MEDIUM,		/* 100 */
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 105 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 110 */
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,		/* 115 */
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_LONG,		/* 120 */
+	TPM_LONG,
+	TPM_MEDIUM,
+	TPM_UNDEFINED,
+	TPM_SHORT,
+	TPM_SHORT,		/* 125 */
+	TPM_SHORT,
+	TPM_LONG,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,		/* 130 */
+	TPM_MEDIUM,
+	TPM_UNDEFINED,
+	TPM_SHORT,
+	TPM_MEDIUM,
+	TPM_UNDEFINED,		/* 135 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 140 */
+	TPM_SHORT,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 145 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 150 */
+	TPM_MEDIUM,
+	TPM_MEDIUM,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_UNDEFINED,		/* 155 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 160 */
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 165 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_LONG,		/* 170 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 175 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_MEDIUM,		/* 180 */
+	TPM_SHORT,
+	TPM_MEDIUM,
+	TPM_MEDIUM,
+	TPM_MEDIUM,
+	TPM_MEDIUM,		/* 185 */
+	TPM_SHORT,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 190 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 195 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 200 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,
+	TPM_SHORT,		/* 205 */
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_MEDIUM,		/* 210 */
+	TPM_UNDEFINED,
+	TPM_MEDIUM,
+	TPM_MEDIUM,
+	TPM_MEDIUM,
+	TPM_UNDEFINED,		/* 215 */
+	TPM_MEDIUM,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,
+	TPM_SHORT,		/* 220 */
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_SHORT,
+	TPM_UNDEFINED,		/* 225 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 230 */
+	TPM_LONG,
+	TPM_MEDIUM,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,		/* 235 */
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_UNDEFINED,
+	TPM_SHORT,		/* 240 */
+	TPM_UNDEFINED,
+	TPM_MEDIUM,
+};
+
+#endif
diff --git a/arch/greco/common/common.c b/arch/greco/common/common.c
index bd46b42..d0c180a 100644
--- a/arch/greco/common/common.c
+++ b/arch/greco/common/common.c
@@ -154,6 +154,11 @@
 	*regs = mmPSOC_QSPI_CTRLR0;
 }
 
+void spi_tpm_base_get(unsigned int *regs)
+{
+	*regs = mmPSOC_SPI_CTRLR0;
+}
+
 /*
  * Set ARC in running state
  * 1. Set run bit in RUN_HALT_REQ register
